### PR TITLE
userkey search uses regex to find field

### PR DESCRIPTION
**Pull request by t0fik**
- regex is used to search field with user name
- default userkey value set to: `(user|login)'

### CODE REVIEW

**Comment by ayushnix** [comment]
Can you explain why this change would be helpful to someone? `tessen` already provides an option to set a custom username key using the `TESSEN_USERKEY` variable. If someone wants to use `login` as the userkey, they can use that.

Do you want to make `tessen` assume that the userkey can be both `user` and `login`?

**Comment by t0fik** [comment]
> Can you explain why this change would be helpful to someone? `tessen` already provides an option to set a custom username key using the `TESSEN_USERKEY` variable. If someone wants to use `login` as the userkey, they can use that.
Some password databases might be inconsistent with user name field name. This change allow to set multiple patterns for userkey.
> 
> Do you want to make `tessen` assume that the userkey can be both `user` and `login`?
Basically yes, but my idea was to allow inconsistency in password database.

**Comment by ayushnix** [comment]
> Basically yes, but my idea was to allow inconsistency in password database.

I can see why some users might use both `user` and `login` as the username key allowing this inconsistency would also reserve two keywords that someone might want to use for different purposes. If someone has set the value of `user` and `login` in the same file, that would be problematic. Moreover, tessen also supports using the basename of the file as the username.

In my opinion, it's better to expect consistency in the database and allow users to change what they want. For what it's worth, rofi-pass also seems to use [only one key](https://github.com/carnager/rofi-pass/blob/629ad8d73a72d90f531ab6ebbdf78db710e25f2f/rofi-pass#L42) for the username. This also incentivizes `tessen` to do the same for people who might want to migrate from rofi-pass to tessen and expect similar behavior.

**Comment by t0fik** [comment]
If that is the point, I can restore default value to `user` only and code will be working as before but I would allow use of regex instead of comparison.

**Comment by ayushnix** [comment]
> If that is the point, I can restore default value to `user` only and code will be working as before but I would allow use of regex instead of comparison.

You mean you'd write something like this?

```sh
tsn_userkey="${TESSEN_USERKEY:-user}"
```
```sh
elif [[ "${key,,}" =~ ^${tsn_userkey}$ ]]; then
```

**Comment by t0fik** [comment]
Exactly.
This way `tessen` will expect consistent naming scheme in password store and user will be able to add fields when needed.

**Comment by ayushnix** [comment]
> Exactly. This way `tessen` will expect consistent naming scheme in password store and user will be able to add fields when needed.

Maybe I'm missing something but what would be the difference between

```sh
elif [[ "${key,,}" =~ ^${tsn_userkey}$ ]]; then
```
 and
```sh
elif [[ "${key,,}" == "$tsn_userkey" ]]; then
```
?

**Comment by t0fik** [comment]
Password store entry:
```
VeryStrongPassword
user: foo
login: bar
username: foobar
...
```

User will be able to configure regular expression to match `userkey` field name, eg: `TESSEN_USERKEY="(user|login|username)"` and then `tessen` will try to match against this regex. First matched field will be taken as `tsn_username` to copy or autotype.
So in this example autotyped values would be:
tsn_username="foobar" (value of `username`)
tsn_password="VeryStrongPassword"

With string comparison test the same value will be taken literally and `tessen` will search field named "(user|login|username)", result would be:
```
tsn_username=filename
tsn_password="VeryStrongPassword"
```

**Comment by ayushnix** [comment]
Got it, sounds good to me. By default, `tessen` should use only a single key but the user can use a regex if he wants to.

**Comment by t0fik** [comment]
Yes. I only need to update man page before ready to merge.

**Comment by ayushnix** [comment]
It's okay, I'll update the man page. I'll review the changes in this PR and #10 tomorrow. Thanks for contributing.

**Review by ayushnix — CHANGES_REQUESTED** [review]

File: tessen [hunk 1]
```diff
@@ -22,7 +22,7 @@ readonly -a tsn_known_backends=('bemenu' 'fuzzel' 'rofi' 'wofi')
 tsn_backend="${TESSEN_BACKEND-}"
 tsn_backend_opts=()
 tsn_action="${TESSEN_ACTION-}"
-tsn_userkey="${TESSEN_USERKEY:-user}"
+tsn_userkey="${TESSEN_USERKEY:-"user"}"
```
[the remark this answers is not in the document]
Comment (t0fik) [reply tessen]: It's not required. I've removed quotes.

File: tessen [hunk 2]
```diff
@@ -85,7 +85,7 @@ get_pass_data() {
     # keys with the case insensitive name 'password' are ignored
     if [[ "${key,,}" == "password" ]]; then
       continue
-    elif [[ "${key,,}" == "$tsn_userkey" ]]; then
+    elif [[ -z "${tsn_username}" ]] && [[ "${key,,}" =~ ^${tsn_userkey}$ ]]; then
       tsn_username="$val"
```
[the remark this answers is not in the document]
Comment (t0fik) [reply tessen:90]: `tsn_userkey` is set during parsing pass file.

**Comment by ayushnix** [comment]
Looks good to me.

**Comment by ayushnix** [comment]
Thanks, this PR made me realize that I should use the value of the first matching key instead of the last matching key in other cases as well. I'll work on that and improve `tessen` further.

**Pull request merged**